From 49eb6e08e831f4c4de77756e23207081dfe16197 Mon Sep 17 00:00:00 2001
From: xubaikun <xubaikun@canaan-creative.com>
Date: Tue, 27 Sep 2022 18:32:34 +0800
Subject: [PATCH] extended gpio_keys driver to supports event operations

---
 drivers/input/keyboard/gpio_keys.c | 99 ++++++++++++++++++++----------
 1 file changed, 67 insertions(+), 32 deletions(-)

diff --git a/drivers/input/keyboard/gpio_keys.c b/drivers/input/keyboard/gpio_keys.c
index 052e3767..5de4580f 100644
--- a/drivers/input/keyboard/gpio_keys.c
+++ b/drivers/input/keyboard/gpio_keys.c
@@ -543,15 +543,16 @@ static int gpio_keys_setup_key(struct platform_device *pdev,
 
 	if (bdata->gpiod) {
 		bool active_low = gpiod_is_active_low(bdata->gpiod);
-
-		if (button->debounce_interval) {
-			error = gpiod_set_debounce(bdata->gpiod,
-					button->debounce_interval * 1000);
-			/* use timer if gpiolib doesn't provide debounce */
-			if (error < 0)
-				bdata->software_debounce =
-						button->debounce_interval;
-		}
+		if (button->type == EV_KEY || button->type == EV_SW)
+		{
+			if (button->debounce_interval) {
+				error = gpiod_set_debounce(bdata->gpiod,
+						button->debounce_interval * 1000);
+				/* use timer if gpiolib doesn't provide debounce */
+				if (error < 0)
+					bdata->software_debounce =
+							button->debounce_interval;
+			}
 
 		if (button->irq) {
 			bdata->irq = button->irq;
@@ -589,6 +590,13 @@ static int gpio_keys_setup_key(struct platform_device *pdev,
 			 * not reconfigure the trigger type.
 			 */
 			break;
+			}
+		}
+		else
+		{
+			error = gpiod_direction_output(bdata->gpiod, active_low);
+			if (error < 0)
+				return error;
 		}
 	} else {
 		if (!button->irq) {
@@ -619,32 +627,34 @@ static int gpio_keys_setup_key(struct platform_device *pdev,
 	*bdata->code = button->code;
 	input_set_capability(input, button->type ?: EV_KEY, *bdata->code);
 
-	/*
-	 * Install custom action to cancel release timer and
-	 * workqueue item.
-	 */
-	error = devm_add_action(dev, gpio_keys_quiesce_key, bdata);
-	if (error) {
-		dev_err(dev, "failed to register quiesce action, error: %d\n",
-			error);
-		return error;
-	}
+	if (button->type == EV_KEY)
+	{
+		/*
+		* Install custom action to cancel release timer and
+		* workqueue item.
+		*/
+		error = devm_add_action(dev, gpio_keys_quiesce_key, bdata);
+		if (error) {
+			dev_err(dev, "failed to register quiesce action, error: %d\n",
+				error);
+			return error;
+		}
 
-	/*
-	 * If platform has specified that the button can be disabled,
-	 * we don't want it to share the interrupt line.
-	 */
-	if (!button->can_disable)
-		irqflags |= IRQF_SHARED;
+		/*
+		 * If platform has specified that the button can be disabled,
+		 * we don't want it to share the interrupt line.
+		 */
+		if (!button->can_disable)
+			irqflags |= IRQF_SHARED;
 
-	error = devm_request_any_context_irq(dev, bdata->irq, isr, irqflags,
-					     desc, bdata);
-	if (error < 0) {
-		dev_err(dev, "Unable to claim irq %d; error %d\n",
-			bdata->irq, error);
-		return error;
+		error = devm_request_any_context_irq(dev, bdata->irq, isr, irqflags,
+						     desc, bdata);
+		if (error < 0) {
+			dev_err(dev, "Unable to claim irq %d; error %d\n",
+				bdata->irq, error);
+			return error;
+		}
 	}
-
 	return 0;
 }
 
@@ -688,6 +698,30 @@ static void gpio_keys_close(struct input_dev *input)
 		pdata->disable(input->dev.parent);
 }
 
+static int gpio_keys_event(struct input_dev *input, unsigned int type, unsigned int code, int value)
+{
+	printk("gpio event, type: 0x%02x, code: 0x%02x, value: 0x%02x\n", type, code, value);
+	struct gpio_keys_drvdata *ddata = input_get_drvdata(input);
+	const struct gpio_keys_platform_data *pdata = ddata->pdata;
+	struct gpio_button_data *bdata = NULL;
+	struct gpio_keys_button *__buttons = NULL;
+	int i = 0;
+	int ret = 0;
+
+	for (i = 0; i < pdata->nbuttons; i++)
+	{
+		__buttons = pdata->buttons + i;
+		bdata = &ddata->data[i];
+		if ((__buttons->type == type) && (__buttons->code == code))
+		{
+			gpiod_set_value(bdata->gpiod, value);
+			break;
+		}
+	}
+
+	return 0;
+}
+
 /*
  * Handlers for alternative sources of platform_data
  */
@@ -816,6 +850,7 @@ static int gpio_keys_probe(struct platform_device *pdev)
 	input->dev.parent = dev;
 	input->open = gpio_keys_open;
 	input->close = gpio_keys_close;
+	input->event = gpio_keys_event;
 
 	input->id.bustype = BUS_HOST;
 	input->id.vendor = 0x0001;
-- 
2.17.1

